app/src/main/java/com/kickstarter/viewmodels/LoginToutViewModel.kt

```diff
@@ -0,0 +1,297 @@
+import com.kickstarter.services.apiresponses.ErrorEnvelope
+import com.kickstarter.ui.IntentKey
+import com.kickstarter.ui.activities.DisclaimerItems
+import com.kickstarter.ui.activities.LoginToutActivity
+import com.kickstarter.ui.data.ActivityResult
+import com.kickstarter.ui.data.LoginReason
+import rx.Notification
+import rx.Observable
+import rx.subjects.BehaviorSubject
+import rx.subjects.PublishSubject
+
+interface LoginToutViewModel {
+    interface Inputs {
+        /** Call when the Login to Facebook button is clicked.  */
+        fun facebookLoginClick(activity: LoginToutActivity?, facebookPermissions: List<String>)
+
+        /** Call when the login button is clicked.  */
+        fun loginClick()
+
+        /** Call when the signup button is clicked.  */
+        fun signupClick()
+
+        /** Call when the disclaimer Item  is clicked.  */
+        fun disclaimerItemClicked(disclaimerItem: DisclaimerItems)
+    }
+
+    interface Outputs {
+        /** Emits when a user has successfully logged in; the login flow should finish with a result indicating success.  */
+        fun finishWithSuccessfulResult(): Observable<Void>
+
+        /** Emits when a user has failed to authenticate using Facebook.  */
+        fun showFacebookAuthorizationErrorDialog(): Observable<String>
+
+        /** Emits when the API was unable to create a new Facebook user.  */
+        fun showFacebookInvalidAccessTokenErrorToast(): Observable<String?>
+
+        /** Emits when the API could not retrieve an email for the Facebook user.  */
+        fun showMissingFacebookEmailErrorToast(): Observable<String?>
+
+        /** Emits when a login attempt is unauthorized.  */
+        fun showUnauthorizedErrorDialog(): Observable<String>
+
+        /** Emits a Facebook user and an access token string to confirm Facebook signup.  */
+        fun startFacebookConfirmationActivity(): Observable<Pair<ErrorEnvelope.FacebookUser, String>>
+
+        /** Emits when the login activity should be started.  */
+        fun startLoginActivity(): Observable<Void>
+
+        /** Emits when the signup activity should be started.  */
+        fun startSignupActivity(): Observable<Void>
+
+        /** Emits when a user has successfully logged in using Facebook, but has require two-factor authentication enabled.  */
+        fun startTwoFactorChallenge(): Observable<Void>
+
+        /** Emits when click one of disclaimer items  */
+        fun showDisclaimerActivity(): Observable<DisclaimerItems>
+    }
+
+    class ViewModel(environment: Environment) :
+        ActivityViewModel<LoginToutActivity?>(environment),
+        Inputs,
+        Outputs {
+        private var callbackManager: CallbackManager? = null
+        private val currentUser: CurrentUserType
+        private val client: ApiClientType
+
+        private fun clearFacebookSession(e: FacebookException) {
+            LoginManager.getInstance().logOut()
+        }
```

Comment:
Noticed there are a bunch of codecov warnings here, should these be covered by tests? Is this a codecov error related to the other issues we've been having with it?

Reply:
For being able to properly test those callbacks we should've created a wrapper client around FacebookSDK (Dependency Inversion Principle, as we do for Apollo, Retrofit, or Braze ->  [DIP](https://dev.to/tamerlang/understanding-solid-principles-dependency-inversion-1b0f#:~:text=The%20Dependency%20Inversion%20Principle%20(DIP,Details%20should%20depend%20upon%20abstractions.) ), that implementation it's out of the scope of this task for sure.
FYI I think there were some conversations around deprecating Facebook as login method, but nothing conclusive yet, so it's ok for now to not have coverage on those areas yet.